util/file.py

```diff
@@ -830,3 +831,129 @@ def join_paired_fastq(input_fastqs, output_format='fastq', num_n=None):
+            out_compress_ps = subprocess.Popen(choose_compressor(pipe_hint)["compress_cmd"], stdout=None if out_compressed_tarball == "-" else outfile, stdin=subprocess.PIPE)
+
+        tar_out = tarfile.open(fileobj=out_compress_ps.stdin, mode="w|")
+
+        for in_compressed_tarball in input_compressed_tarballs:
+            if in_compressed_tarball != "-":
+                pigz_ps = subprocess.Popen(choose_compressor(in_compressed_tarball)["decompress_cmd"] + [in_compressed_tarball], stdout=subprocess.PIPE)
+            else:
+                assert in_compressed_tarball != '-' or pipe_hint, "cannot autodetect compression on stdin input unless pipeHint provided"
+                pigz_ps = subprocess.Popen(choose_compressor(pipe_hint)["decompress_cmd"] + [in_compressed_tarball], stdout=subprocess.PIPE, stdin=sys.stdin)
+            tar_in = tarfile.open(fileobj=pigz_ps.stdout, mode="r|", ignore_zeros=True)
+
+            fileinfo = tar_in.next()
+            while fileinfo is not None:
+                if extract_to_disk_path:
+                    target_path = os.path.normpath(os.path.join(extract_to_disk_path, fileinfo.name).rstrip("/"))
+                    containing_path = os.path.dirname(target_path)
+                    create_containing_dirs(containing_path)
+
+                    if avoid_disk_roundtrip:
+                        fileobj = tar_in.extractfile(fileinfo)
+                        #tar_out.addfile(fileinfo)
+
+                        # if it is a file, write it out
+                        if fileinfo.isreg():
+                            tar_out.addfile(fileinfo, fileobj=FileDiverter(fileinfo, fileobj, written_mirror_file=target_path))
+                        # if this is not a file (dir, symlink, etc.) use the disk roundtrip since it's fast enough
+                        else:
+                            outfile = tar_in.extract(fileinfo, path=extract_to_disk_path)
+                            with pushd_popd(extract_to_disk_path):
+                                tar_out.add(fileinfo.name)
+                        
+                    else:
+                        # alternative implementation:
+                        outfile = tar_in.extract(fileinfo, path=extract_to_disk_path)
+                        with pushd_popd(extract_to_disk_path):
+                            tar_out.add(fileinfo.name)
+                else:
+                    fileobj = tar_in.extractfile(fileinfo)
+                    tar_out.addfile(fileinfo, fileobj=fileobj)
+
+                
+                fileinfo = tar_in.next()
+            pigz_ps.stdout.flush()
+            pigz_ps.wait()
+            tar_in.close()
+            if pigz_ps.poll():
+                raise subprocess.CalledProcessError(pigz_ps.returncode, "Call error %s" % pigz_ps.returncode)
+
+        tar_out.close()
+        out_compress_ps.stdin.flush()
+        out_compress_ps.stdin.close()
+        out_compress_ps.wait()
+        if out_compress_ps.poll():
```

Comment:
Should this be `if ps.returncode != 0`?

Reply:
True, but it works though... `Popen.poll()` returns either the returncode (if it's done) or `None` (if it's not done, but it's guaranteed to be done after the `.wait()` anyway). I think we have a few other places like this throughout our code that could be converted for clarity, but would likely not change behavior.